_posts/2017-04-20-how-software-works-part1.md

```diff
@@ -0,0 +1,196 @@
+                                     42
+```
+
+Once you know how to convert between number bases, you realise that binary 00101010
+and decimal 42 are _the same thing_. They express exactly the same number just like
+1km expresses the exact same distance as .621 miles.
+
+There are some limits, however, that we must consider. Just like a 3-digit decimal number
+can only express numbers between 0-999, there are constraints to binary numbers. For example,
+an 8 bit binary number can express at most
+
+`2^8 = 256`
+
+discrete values.
+
+Our current representation has another limitation: it can only express positive integers.
+We need to add a little twist to be able to represent negative integers. The change is
+to reserve the left-most bit (also called the most significant bit) to signal whether the
+number is positive or negative. For example:
+
+```
+00000001 == 2
+```
+
+```
+10000001 == -2
+```
+
+This isn't a magic trick, however. An 8 bit number can still only represent 256 discrete
+values, the range shifts from `0...256` to `-128...128`.
+
+Both representations are used often in programming. They're called _signed_ and _unsigned_
+integers depending on whether the most significant bit is treated as a sign bit or not.
+Choosing between the two is a tradeoff – signed numbers can go negative but the maximum positive number drops to half.
+
+I didn't choose the number 8 randomly: a string of 8 bits forms what we call a _byte_.
+While bytes are small and sufficient for a surprisingly large number of tasks,
+they are still very limited in terms of size. 256 numbers is not nearly enough to
+run complex calculations or show, for example, a bank account balance. On modern computers,
+you can choose to use more than 8 bits to represent an integers for these purposes. This is
+called the integer _width_. Computers are best at dealing with integers whose width is a power of
+two, so most integers used in the wild have either 8, 16, 32, 64, or 128 bit width.
+
+Good. We have a way to encode human-friendly numbers to a format that our hardware can work with.
+You might be asking at this point how useful it is to just being able to store numbers in binary.
+Surely the data that we need to work with these days is more complex than that.
+Hold that thought while we dive into how integers are used to encode the lowest-level form
+of software: Machine code.
+
+### 1 10 20 0
+
+From this point on, we will assume that a device exists which can run machine code.
+It has a CPU, RAM memory, buses to transfer data etc. Generally what one would call a computer
+architecture. For example, I'm writing this blog on a MacBook Pro which has one of the most common
+computer architectures called x86. The architecture has intrinsic, hardware-level routines to
+do very simple computations, move data between RAM and the CPU, manipulate registers etc. We will
+get to what all of that means but for now let's just assume that the hardware is there.
```

Comment:
We will get to what that means later. For now, let's assume the hardware is available.